maintainers/scripts/haskell/hydra-report.hs

```diff
@@ -0,0 +1,233 @@
+
+newtype JobsetEvalInputs = JobsetEvalInputs {nixpkgs :: Nixpkgs}
+   deriving (Generic, ToJSON, FromJSON, Show)
+
+data Eval = Eval
+   { builds :: Seq Int
+   , id :: Int
+   , jobsetevalinputs :: JobsetEvalInputs
+   }
+   deriving (Generic, ToJSON, FromJSON, Show)
+
+data Build = Build
+   { job :: Text
+   , buildstatus :: Int
+   , finished :: Int
+   , id :: Int
+   , nixname :: Text
+   , system :: Text
+   , jobsetevals :: Seq Int
+   }
+   deriving (Generic, ToJSON, FromJSON, Show)
+
+data BuildName = BuildName Text Int deriving (Show, Eq, Ord)
+
+data StatusSummary = StatusSummary
+   { successfulBuilds :: Int
+   , unfinishedBuilds :: Int
+   , canceledBuilds :: Int
+   , failedMaintainedBuilds :: Seq (BuildName, Seq Text)
+   , failedMaintainedDependencies :: Seq (BuildName, Seq Text)
+   , failedPackageBuilds :: Seq BuildName
+   , failedPackageDependencies :: Seq BuildName
+   , failedJobBuilds :: Seq BuildName
+   , failedJobDependencies :: Seq BuildName
+   }
+   deriving (Show)
+
+instance Semigroup StatusSummary where
+   (<>) lhs rhs =
+      StatusSummary
+         { successfulBuilds = successfulBuilds lhs + successfulBuilds rhs
+         , failedPackageBuilds = failedPackageBuilds lhs <> failedPackageBuilds rhs
+         , failedPackageDependencies = failedPackageDependencies lhs <> failedPackageDependencies rhs
+         , failedJobBuilds = failedJobBuilds lhs <> failedJobBuilds rhs
+         , failedJobDependencies = failedJobDependencies lhs <> failedJobDependencies rhs
+         , failedMaintainedBuilds = failedMaintainedBuilds lhs <> failedMaintainedBuilds rhs
+         , failedMaintainedDependencies = failedMaintainedDependencies lhs <> failedMaintainedDependencies rhs
+         , unfinishedBuilds = unfinishedBuilds lhs + unfinishedBuilds rhs
+         , canceledBuilds = canceledBuilds lhs + canceledBuilds rhs
+         }
+
+instance Monoid StatusSummary where
+   mempty = StatusSummary 0 0 0 mempty mempty mempty mempty mempty mempty
+
+main :: IO ()
+main = do
+   args <- getArgs
+   case args of
+      ["get-report"] -> getBuildReports
+      ["ping-maintainers"] -> printMaintainerPing
+      ["mark-broken-list"] -> printMarkBrokenList
+      _ -> putStrLn "Usage: get-report | ping-maintainers | mark-broken-list"
+
+reportFileName = "haskell-updates-build-report.json"
+
+showT :: Show a => a -> Text
+showT = Text.pack . show
+
+getBuildReports :: IO ()
+getBuildReports = runReq defaultHttpConfig $ do
+   -- GET request http response
+   now <- liftIO getCurrentTime
+   r <- req GET (https "hydra.nixos.org" /: "jobset" /: "nixpkgs" /: "haskell-updates" /: "evals") NoReqBody jsonResponse mempty
+   let eval = Seq.lookup 0 . evals $ (responseBody r :: JobsetEvals)
+   eval & maybe (liftIO $ putStrLn "No Evalution found") \eval -> do
+      (buildReports :: Seq Build) <- forM (builds eval) \buildId ->
+         responseBody <$> req GET (https "hydra.nixos.org" /: "build" /: showT buildId) NoReqBody jsonResponse mempty
```

Comment:
I think this should retry on failure a few times. Hydra has the habit of randomly 500ing on you.

Reply:
Interestingly this has never failed me until now.

Actually I realize now, what we want is to keep one connection open and query via that one. I have no clue if req does that. Need to look into that.

But retrying is certainly a good idea.

Reply:
Okay. I have read the documentation of `req` and apparently it just has really good defaults. It already does connection sharing and retrying under the hood.